csrc/selective_scan/selective_scan_bwd_kernel.cuh

```diff
@@ -499,14 +505,18 @@ void selective_scan_bwd_launch(SSMParamsBwd &params, cudaStream_t stream) {
                         // using Ktraits = Selective_Scan_bwd_kernel_traits<kNThreads, kNItems, true, kIsVariableB, kIsVariableC, kDeltaSoftplus, kHasZ, input_t, weight_t>;
                         // TODO: check this
                         constexpr int kSmemSize = Ktraits::kSmemSize + MAX_DSTATE * sizeof(typename Ktraits::scan_t) + (kNThreads + 4 * MAX_DSTATE) * sizeof(typename Ktraits::weight_t);
-                        // printf("smem_size = %d\n", kSmemSize);
+
                         dim3 grid(params.batch, params.dim);
-                        auto kernel = &selective_scan_bwd_kernel<Ktraits>;
+                        const void * kernel = (void*) &selective_scan_bwd_kernel<Ktraits>; // TODO: change to reinterpret cast? what's the best practice?
+
                         if (kSmemSize >= 48 * 1024) {
                             C10_CUDA_CHECK(cudaFuncSetAttribute(
                                 kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, kSmemSize));
-                        }
-                        kernel<<<grid, Ktraits::kNThreads, kSmemSize, stream>>>(params);
+                        }  
+
+                        auto kernel_fn = (void (*const)(SSMParamsBwd)) kernel; // Todo - double-check. Had to add this C-style conversion. // TODO: change to reinterpret cast?
+
+                        kernel_fn<<<grid, Ktraits::kNThreads, kSmemSize, stream>>>(params);
```

Comment:
Take this comment as optional.

Note that hipFuncSetAttribute() is a no-op on HIP.  HIP doesn't have a dynamic shared mem size.  You could #ifdef away all this and leave line 509 the kernel launch alone.  

```C++
#ifndef USE_ROCM
auto kernel = &selective_scan_bwd_kernel<Ktraits>;
if (kSmemSize >= 48 * 1024) {
    C10_CUDA_CHECK(cudaFuncSetAttribute(
        kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, kSmemSize));
}
kernel<<<grid, Ktraits::kNThreads, kSmemSize, stream>>>(params);
#else
selective_scan_bwd_kernel<Ktraits><<<grid, Ktraits::kNThreads, kSmemSize, stream>>>(params);
#endif
```

Then again, it would be nice for hipify to do a sensible thing here and just update the code to be compatible with hip.  If you plan to upstream this code, you might just need the ifdef above so that the cuda code path is not affected.

Reply:
Do you think we might want to raise some sort of error if we reach this code that attempts to adjust shared memory size? I.e. I'm a bit worried we might silently hit a shared memory size limit. Or, more generally, should we re-adjust the way we currently set the shared memory size in HIP to make sure we allocate enough (given that we don't have the luxury of dynamically readjusting it)?

Reply:
I confirmed with someone who knows HIP well.  If you try to launch with a shared mem size that is above what the hardware supports, you will trigger a failure at runtime.  For a better user experience, we should favor no-ops over raising errors, and for hardware-specific features we have no hope of porting in software, we should not support such APIs or attributes.  This dynamic shmem size is kindof a gray area.  It was less friction to hipify it even though it doesn't nothing, which might be better than a compile-time failure or runtime failure.